File: compiler.py
import sys

# Conversion Functions
def num_to_ba(num: str):
    hnum = hex(int(num)).replace("0x","")
    rba = "0"*(4-len(hnum))+hnum
    ba = [rba[i:i + 2] for i in range(0, len(rba), 2)]
    iba = [int(i,16) for i in ba]
    return iba

def str_to_hex(s: str) -> str:
    return s.encode("utf-8").hex()

def str_to_ba(s):
    s = str_to_hex(s)
    return [int(s[i:i + 2],16) for i in range(0, len(s), 2)]

def op_to_by(op: str) -> int:
    if op == "+": return 1
    elif op == "-": return 2
    elif op == "*": return 3
    elif op == "/": return 4
    elif op == ">": return 10
    elif op == "<": return 11
    elif op == ">=": return 12
    elif op == "<=": return 13
    elif op == "!=": return 14
    elif op == "==": return 15
    else: return 0

def bool_to_by(bool: str) -> int:
    if bool == "true": return 2
    elif bool == "false": return 1
    else: return 0

input_file = open("test.asm","r").read()
content = input_file.split("\n")

header = [1,0,64,0,64,0,0,0,0,0,0,0,0,0,0,0]
inst = []

# Flags to check if code is proper
ret_exists = False
ret_correct = False

header_exists = False
header_correct = False

# Translating the code
cur_line = 1
for line in content:
    # TODO: REPLACE WITH SWITCH
    if line.lower().startswith("header:"): # Header
        header_exists = True
        head = line.split(" ")
        head[0] = head[0][7:].lower()

        if head[1] == "terminal": header[0] = 1
        elif head[1] == "graphical": header[0] = 0

        header[1:3] = num_to_ba(head[2])
        header[3:5] = num_to_ba(head[3])

        header_correct = True

        del head

    elif line.lower().startswith(";"): # Comment
        pass

    elif line.lower().startswith("nop"): # 0x00 Inst; NOP
        inst.extend([0])

    elif line.lower().startswith("lbl"): # 0x01 Inst; LBL String hello "Hello World!"
        con = line.split(" ")
        
        lbl_name = str_to_ba(con[2])

        lbl_type = ""
        if con[1].lower() == "number": lbl_type = 3
        elif con[1].lower() == "string": lbl_type = 2
        elif con[1].lower() == "boolean": lbl_type = 1
        else: raise Exception("COMPILER::USER_CODE::UNKNOWN_LABEL_TYPE")

        lbl_con = ' '.join(con[3:])
        lbl_content = str_to_ba(lbl_con)
        if lbl_type == 3: 
            lbl_content = num_to_ba(lbl_con)
            inst.extend([1,lbl_type]+lbl_name+[0, lbl_content[0], lbl_content[1]])
        else:
            inst.extend([1,lbl_type]+lbl_name+[0]+lbl_content)

        del lbl_name, lbl_con, lbl_content, lbl_type, con

    elif line.lower().startswith("mth"): # 0x02 Inst; MTH num1 + 10
        con = line.split(" ")
        if con[1].isdigit(): raise Exception("COMPILER::USER_CODE::LABEL_ARG_IS_NUMERIC\nLine: "+str(cur_line))

        num1 = str_to_ba(con[1])
        num2 = str_to_ba(con[3])
        num_type = 6

        if con[3].isdigit():
            num2 = num_to_ba(con[3])
            num_type = 3

        type = op_to_by(con[2])

        inst.extend([2,type]+num1+[0,num_type]+num2+[0])

        del con, num1, num2, type

    elif line.lower().startswith("put"): # 0x03 Inst; PUT hello; PUT "Hello World!"
        con = line.split(" ")

        type = 3
        if "\"" in con[1]: type = 2
        elif con[1] == "true" or con[1] == "false": type = 1
        elif con[1].startswith("#"): type = 4
        elif not con[1].isdigit(): type = 6

        put_content = 0
        if type == 2: put_content = str_to_ba(" ".join(con[1:])) # String
        elif type == 3: put_content = num_to_ba(con[1]) # Number
        elif type == 4: put_content = str_to_ba(con[1].replace("#","")) # Color
        elif type == 1 or type == 6: put_content = str_to_ba(con[1]) # Label / Boolean

        inst.extend([3,type]+put_content+[0])

        del con, type

    elif line.lower().startswith("rki"): # 0x04 Inst; RKI yourname
        con = line.split(" ")

        if len(con) > 2: raise Exception("COMPILER::USER_CODE::RKI_TOO_LONG\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].isdigit(): raise Exception("COMPILER::USER_CODE::LABEL_IS_NUMBER\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].startswith("\""): raise Exception("COMPILER::USER_CODE::LABEL_IS_STRING\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))

        inst.extend([4]+str_to_ba(con[1])+[0])

        del con

    elif line.lower().startswith("ret"): # 0x05 Inst; RET 24; RET deez
        ret_exists = True
        con = line.split(" ")

        type = 3 if con[1].isdigit() else 6
        ret_code = num_to_ba(con[1]) if type == 3 else str_to_ba(con[1])

        inst.extend([5,type]+ret_code+[0])
        ret_correct = True

        del con, type

    elif line.lower().startswith("mrk"): # 0x06 Inst; MRK Slime
        con = line.split(" ")

        if len(con) > 2: raise Exception("COMPILER::USER_CODE::MRK_TOO_LONG\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].isdigit(): raise Exception("COMPILER::USER_CODE::MARKER_IS_NUMBER\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].startswith("\""): raise Exception("COMPILER::USER_CODE::MARKER_IS_STRING\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))

        inst.extend([6]+str_to_ba(con[1])+[0])

        del con

    elif line.lower().startswith("jmp"): # 0x07 Inst; JMP Slime
        con = line.split(" ")

        if len(con) > 2: raise Exception("COMPILER::USER_CODE::JMP_TOO_LONG\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].isdigit(): raise Exception("COMPILER::USER_CODE::MARKER_IS_NUMBER\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))
        if con[1].startswith("\""): raise Exception("COMPILER::USER_CODE::MARKER_IS_STRING\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))

        inst.extend([7]+str_to_ba(con[1])+[0])

        del con

    elif line.lower().startswith("ssc"): # 0x08 Inst; SSC 800 800
        con = line.split(" ")

        num1t, num2t, num1, num2 = 0, 0, 0, 0

        if con[1].isdigit():
            num1t = 3
            num1 = num_to_ba(con[1])

        if con[2].isdigit():
            num2t = 3
            num2 = num_to_ba(con[2])

        inst.extend([8,num1t]+num1+[0,num2t]+num2+[0])

        del con, num1t, num2t, num1, num2

    elif line.lower().startswith("bep"): # 0x09 Inst; BEP 1000 1000
        con = line.split(" ")

        num1t, num2t, num1, num2 = 0, 0, 0, 0

        if con[1].isdigit():
            num1t = 3
            num1 = num_to_ba(con[1])

        if con[2].isdigit():
            num2t = 3
            num2 = num_to_ba(con[2])

        inst.extend([9,num1t]+num1+[0,num2t]+num2+[0])

        del con, num1t, num2t, num1, num2

    elif line.lower().startswith("ifj"): # 0x0A Inst; IFJ 10 >= num1 marker01
        con = line.split(" ")

        num1t, num2t, num1, num2, op, marker = 6, 6, str_to_ba(con[1]), str_to_ba(con[3]), op_to_by(con[2]), str_to_ba(con[4])

        if con[1].isdigit():
            num1t = 3
            num1 = num_to_ba(con[1])
        elif con[1] == "true" or con[1] == "false":
            num1t = 1
            num1 = bool_to_by(con[1])

        if con[3].isdigit():
            num2t = 3
            num2 = num_to_ba(con[3])
        elif con[3] == "true" or con[3] == "false":
            num2t = 1
            num2 = bool_to_by(con[3])

        inst.extend([10,op,num1t]+num1+[0,num2t]+num2+[0]+marker+[0])

        del con, num1t, num2t, num1, num2, op, marker

    elif line.lower().startswith("dtb"): # 0x0B Inst; DTB 300 250 #0495AB | 0B 03 021C 00 03 00FA 00 04 0495AB 00
        con = line.split(" ")

        num1t, num2t, num1, num2 = 0, 0, 0, 0

        type = 4 if con[1].startswith("#") else 6
        num = str_to_ba(con[2].replace("#","")) if type == 4 else str_to_ba(con[2])

        if con[1].isdigit():
            num1t = 3
            num1 = num_to_ba(con[1])

        if con[2].isdigit():
            num2t = 3
            num2 = num_to_ba(con[2])

        inst.extend([11,num1t]+num1+[0,num2t]+num2+[0,type]+num)

        del con, num1t, num2t, num1, num2, type, num

    elif line.lower().startswith("cdb"): # 0x0C Inst; CDB #000000
        con = line.split(" ")

        type = 4 if con[1].startswith("#") else 6
        num = str_to_ba(con[2].replace("#","")) if type == 4 else str_to_ba(con[2])

        inst.insert([12+type]+num+[0])

        del con, type, num

    elif line.lower().startswith("rfb"): # 0x0D Inst; RFB 301 250
        con = line.split(" ")

        num1t, num2t, num1, num2 = 0, 0, 0, 0

        if con[1].isdigit():
            num1t = 3
            num1 = num_to_ba(con[1])

        if con[2].isdigit():
            num2t = 3
            num2 = num_to_ba(con[2])

        inst.extend([13,num1t]+num1+[0,num2t]+num2+[0])

        del con, num1, num2, num1t, num2t

    elif line.lower().startswith("wft"): # 0x0F Inst; WFT 1000
        con = line.split(" ")

        type = 3 if con[1].isdigit() else 6
        num = num_to_ba(con[2]) if con[1].isdigit() else str_to_ba(con[2])

        inst.extend([14,type]+num+[0])

        del con, type, num

    elif str.isspace(line) or line == '':
        pass

    else: raise Exception("COMPILER::USER_CODE::UNKNOWN_INST\nLine: "+str(cur_line)+"\nFile: "+str(sys.argv[0]))

    cur_line += 1

del cur_line, content, input_file

# Raising Exceptions if user forgot basic shit
if not ret_exists: raise Exception("COMPILER::USER_CODE::RETURN_MISSING")
if not ret_correct: raise Exception("COMPILER::USER_CODE::RETURN_MALFORMED")

if not header_exists: raise Exception("COMPILER::USER_CODE::HEADER_MISSING")
if not header_correct: raise Exception("COMPILER::USER_CODE::HEADER_MALFORMED")

# Converting into binary
bin_content = bytearray()
bin_content.extend(header)
bin_content.extend(inst)

# Writing Binary to File
bin_file = open("test.bin","wb")
bin_file.write(bin_content)
bin_file.close()

File: Compiler.cs
﻿using System.Globalization;
using System.Text;

namespace COBE_CS {
    internal class Compiler {
        string prefix = "";
        string[] asm_content = { };
        public int cur_line = 0;
        string title = "Program";

        List<Byte> header = new List<byte>();
        List<Byte> imps = new List<byte>();
        public List<Byte> inst = new List<byte>();

        bool ret_exists = false;
        bool ret_correct = false;
        bool header_exists = false;
        bool header_correct = false;
        bool custom_title = false;
        bool has_start = false;
        bool impscr = false;
        public bool flag_err = false;

        public void PrintA(string type, char[] bytes) {
            var sb = new StringBuilder(type + " compiled as byte { ");
            int index = 0;
            foreach (var b in bytes) {
                sb.Append(b);
                if (bytes.Length - 1 > index) sb.Append(", ");
                index += 1;
            }
            sb.Append(" }");
            Console.WriteLine(sb.ToString());
        }

        // ------------------- Conversion Functions ---------------------
        public byte[] intToBa(int num) {
            string s = num.ToString("X");
            List<Byte> hex = new List<byte>();

            // "320" -> "0320" -> [ 03, 20 ]
            string zeros = "";
            for (int i = 4 - s.Length; i > 0; i--) {
                zeros += "0";
            }
            s = zeros + s;

            // The number is now 4 long, split it in half.
            hex.Add(Byte.Parse(s[0..2], NumberStyles.HexNumber));
            hex.Add(Byte.Parse(s[2..4], NumberStyles.HexNumber));

            return hex.ToArray();
        }

        public byte[] strToBa(string s) {
            List<Byte> intStr = new List<byte>();

            int position = 0;
            bool next = false;
            foreach (char letter in s) {
                if (next) { next = false; position += 1; continue; }

                int value = Convert.ToInt32(letter);

                if (letter == '\\') {
                    switch (s[position + 1]) {
                        case 't':
                            next = true;
                            value = 9;
                            break;
                        case 'n':
                            next = true;
                            value = 10;
                            break;
                        default:
                            throw new Exception("UNKNOWN_ESCAPE_SEQUENCE: " +
                                letter + s[position + 1]);
                    }

                }

                intStr.Add((byte)value);
                position += 1;
            }
            return intStr.ToArray();
        }

        public byte opToBy(string op) {
            switch (op) {
                case "+": return 1;
                case "-": return 2;
                case "*": return 3;
                case "/": return 4;
                case "=": return 5;
                case ">": return 10;
                case "<": return 11;
                case ">=": return 12;
                case "<=": return 13;
                case "!=": return 14;
                case "==": return 15;
            }
            return 0;
        }

        public byte boolToBy(string b) {
            if (b == "true") return 2;
            else if (b == "false") return 1;
            else return 0;
        }

        public byte[] colToBa(string s) {
            s = s.Replace("#",""); // #FF00FF -> FF00FF
            byte r = Convert.ToByte(s[0..2], 16);
            byte g = Convert.ToByte(s[2..4], 16);
            byte b = Convert.ToByte(s[4..6], 16);
            return [r, g, b]; // FF00FF -> 255,0,255
        }

        // ------------------- Assembly Load ---------------------
        public void LoadASM(string path) {
            asm_content = File.ReadAllLines(path);
        }

        // ------------------- Compilation ---------------------
        public void Compile() {
            foreach (var line in asm_content) {
                string[] con = line.Split(' ');

                switch (con[0].ToLower()) {
                    case "header:": {
                        byte mode = 0;
                        header_exists = true;

                        if (con[1].ToLower() == "terminal") mode = 1;
                        else if (con[1].ToLower() == "graphical") mode = 0;
                        else if (con[1].ToLower() == "library") mode = 2;
                        else throw new Exception("UNKNOWN_MODE");

                        byte[] width = [0, 0];
                        byte[] height = [0, 0];

                        if (mode != 2) {
                            width = intToBa(Int32.Parse(con[2]));
                            height = intToBa(Int32.Parse(con[3]));
                        }

                        header.Add(mode);
                        header.AddRange(width);
                        header.AddRange(height);
                        header.AddRange([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

                        header_correct = true;
                        break;
                    }

                    case "title:": {
                        string s = string.Join(" ", con[1..]).Replace("\"", "");
                        if (s.Length > 32) s = s[..32];

                        List<byte> zeros = new List<byte>();
                        for (int i = 32 - s.Length; i > 0; i--) {
                            zeros.Add(0);
                        }

                        header.AddRange(strToBa(s));
                        header.AddRange(zeros);
                        custom_title = true;
                        break;
                    }

                    case String s when string.IsNullOrWhiteSpace(s) || s.StartsWith(";"): {
                        break;
                    }

                    case "nop": {
                        inst.Add(0);
                        break;
                    }

                    case "lbl": {
                        // Label Type
                        byte lbl_type = 0;
                        string lbl_con = String.Join(" ", con[3..]);
                        byte[] lcon = [];

                        switch (con[1].ToLower()) {
                            case "boolean":
                                lbl_type = 1;
                                lcon = [boolToBy(lbl_con)];
                                break;

                            case "string":
                                lbl_type = 2;
                                lcon = strToBa(lbl_con);
                                break;

                            case "number":
                                lbl_type = 3;
                                lcon = intToBa(int.Parse(lbl_con));
                                break;

                            case "color":
                                lbl_type = 4;
                                lcon = colToBa(lbl_con);
                                break;

                            case "file":
                                break;

                            default:
                                throw new Exception("UNKNOWN_LABEL_TYPE");
                        }

                        List<byte> ninst = [1, lbl_type, ..strToBa(prefix + con[2]), 0];
                        ninst.AddRange(lcon);
                        ninst.Add(0);

                        inst.AddRange(ninst);
                        break;
                    }

                    case "mth": {
                        if (con[1].All(char.IsDigit)) throw new Exception("LABEL_ARG_IS_NUMERIC");

                        byte[] num1 = strToBa(con[1]);
                        byte[] num2 = strToBa(con[3]);
                        byte numt = 6;

                        if (con[3].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[3]));
                            numt = 3;
                        } else if (con[3].Contains("\"")) {
                            num2 = strToBa(string.Join(" ", con[3..]));
                            numt = 2;
                        }
                        byte op = opToBy(con[2]);

                        List<byte> ninst = [2, op, .. num1, 0, numt, .. num2, 0];

                        inst.AddRange(ninst);
                        break;
                    }

                    case "put": {
                        byte type = 6;
                        byte[] pcon = strToBa(con[1]);

                        if (con[1].Contains("\"")) {
                            type = 2;
                            pcon = strToBa(String.Join(" ", con[1..]));
                        } else if (con[1] == "true" || con[1] == "false") {
                            type = 1;
                            pcon = new byte[] { boolToBy(con[1]) };
                        } else if (con[1].StartsWith("#")) {
                            type = 4;
                            pcon = strToBa(con[1].Replace("#", ""));
                        } else if (con[1].All(char.IsDigit)) {
                            type = 3;
                            pcon = intToBa(Int32.Parse(con[1]));
                        }

                        List<byte> ninst = [3, type, .. pcon, 0];

                        inst.AddRange(ninst);
                        break;
                    }

                    case "rki": {
                        if (con.Length > 2) throw new Exception("RKI_TOO_LONG");
                        if (con[1].All(char.IsDigit)) throw new Exception("NUMERIC_AS_LABEL");
                        if (con[1].StartsWith("\"")) throw new Exception("STRING_AS_LABEL");

                        List<byte> ninst = [4, .. strToBa(con[1]), 0];

                        inst.AddRange(ninst);
                        break;
                    }

                    case "ret": {
                        ret_exists = true;

                        byte type = 6;
                        byte[] ret_code = strToBa(con[1]);
                        if (con[1].All(char.IsDigit)) {
                            type = 3;
                            ret_code = intToBa(Int32.Parse(con[1]));
                        }

                        List<byte> ninst = [5, type, ..ret_code, 0];

                        inst.AddRange(ninst);
                        ret_correct = true;
                        break;
                    }

                    case "mrk": {
                        if (con.Length > 2) throw new Exception("MRK_TOO_LONG");
                        if (con[1].All(char.IsDigit)) throw new Exception("NUMERIC_AS_MARKER");
                        if (con[1].StartsWith("\"")) throw new Exception("STRING_AS_MARKER");

                        List<byte> ninst = [];
                        if (con[1] == "_start") {
                            ninst.AddRange([15, 1]);
                            has_start = true;
                        } else ninst.AddRange([6, ..strToBa(con[1]), 0]);

                        inst.AddRange(ninst);
                        break;
                    }

                    case "jmp": {
                        if (con.Length > 2) throw new Exception("JMP_TOO_LONG");
                        if (con[1].All(char.IsDigit)) throw new Exception("NUMERIC_AS_MARKER");
                        if (con[1].StartsWith("\"")) throw new Exception("STRING_AS_MARKER");

                        List<byte> ninst = [7, ..strToBa(con[1]), 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "ssc": {
                        byte[] num1 = strToBa(con[1]), num2 = strToBa(con[2]);
                        byte num1t = 6, num2t = 6;

                        if (con[1].All(char.IsDigit)) {
                            num1 = intToBa(Int32.Parse(con[1]));
                            num1t = 3;
                        }

                        if (con[2].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[2]));
                            num2t = 3;
                        }

                        List<byte> ninst = [8, num1t, ..num1, 0, num2t, .. num2, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "bep": {
                        byte[] num1 = strToBa(con[1]), num2 = strToBa(con[2]);
                        byte num1t = 6, num2t = 6;

                        if (con[1].All(char.IsDigit)) {
                            num1 = intToBa(Int32.Parse(con[1]));
                            num1t = 3;
                        }

                        if (con[2].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[2]));
                            num2t = 3;
                        }

                        List<byte> ninst = [9, num1t, ..num1, 0, num2t, ..num2, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "ifj": {
                        byte[] num1 = strToBa(con[1]), num2 = strToBa(con[3]);
                        byte[] marker = strToBa(con[4]);
                        byte num1t = 6, num2t = 6, op = opToBy(con[2]);

                        if (con[1].All(char.IsDigit)) {
                            num1 = intToBa(Int32.Parse(con[1]));
                            num1t = 3;
                        } else if (con[1] == "true" || con[1] == "false") {
                            num1t = 1;
                            num1 = new byte[] { boolToBy(con[1]) };
                        }

                        if (con[3].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[3]));
                            num2t = 3;
                        } else if (con[1] == "true" || con[1] == "false") {
                            num2t = 1;
                            num2 = new byte[] { boolToBy(con[1]) };
                        }

                        List<byte> ninst = [10, op, num1t, ..num1, 0, num2t, ..num2, 0, ..marker, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "dtb": {
                        byte[] num1 = strToBa(con[1]), num2 = strToBa(con[2]);
                        byte[] col = strToBa(con[3]);
                        byte num1t = 6, num2t = 6, type = 6;

                        if (con[3].StartsWith("#")) {
                            type = 4;
                            col = strToBa(con[2].Replace("#", ""));
                        }

                        if (con[1].All(char.IsDigit)) {
                            num1 = intToBa(Int32.Parse(con[1]));
                            num1t = 3;
                        }

                        if (con[2].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[2]));
                            num2t = 3;
                        }

                        List<byte> ninst = [11, num1t, ..num1, 0, num2t, ..num2, 0, type, ..col, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "cdb": {
                        byte type = 6;
                        byte[] num = strToBa(con[2].Replace("#", ""));

                        List<byte> ninst = [12, type, ..num, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "rfb": {
                        byte[] num1 = strToBa(con[1]), num2 = strToBa(con[2]);
                        byte num1t = 6, num2t = 6;

                        if (con[1].All(char.IsDigit)) {
                            num1 = intToBa(Int32.Parse(con[1]));
                            num1t = 3;
                        }

                        if (con[2].All(char.IsDigit)) {
                            num2 = intToBa(Int32.Parse(con[2]));
                            num2t = 3;
                        }

                        List<byte> ninst = [13, num1t, ..num1, 0, num2t, ..num2, 0];
                        inst.AddRange(ninst);
                        break;
                    }

                    case "wft": {
                        byte[] num = strToBa(con[1]);
                        byte numt = 6;

                        if (con[1].All(char.IsDigit)) {
                            num = intToBa(Int32.Parse(con[1]));
                            numt = 3;
                        }

                        List<byte> ninst = [ 14, numt, ..num, 0];

                        inst.AddRange(ninst);
                        break;
                    }

                    case "imp": {
                        string impscript = con[1].Replace("/", ".");

                        Compiler imp_comp = new Compiler();
                        imp_comp.prefix = impscript + ".";

                        // Attempt to compile the imported script
                        imp_comp.LoadASM(impscript + ".asm");
                        imp_comp.Compile();

                        // Merge with the main script
                        imps.AddRange(imp_comp.inst);
                        impscr = true;
                        break;
                    }

                    case "rtj": {
                        if (con.Length > 2) throw new Exception("MRK_TOO_LONG");
                        if (con[1].All(char.IsDigit)) throw new Exception("NUMERIC_AS_MARKER");
                        if (con[1].StartsWith("\"")) throw new Exception("STRING_AS_MARKER");

                        List<byte> ninst = [17, .. strToBa(con[1]), 0];

                        inst.AddRange(ninst);
                        break;
                    }

                    default:
                        throw new Exception("UNKNOWN_INST: " + con[0].ToLower());
                }

                cur_line++; 
            }
        }

        // Check if everything is in order
        public void CheckFlags() {
            flag_err = true;
            if (!header_exists) throw new Exception("HEADER_NOT_PRESENT");
            if (!header_correct) throw new Exception("HEADER_NOT_CORRECT");

            if (!ret_exists) throw new Exception("RETURN_NOT_PRESENT");
            if (!ret_correct) throw new Exception("RETURN_NOT_PRESENT");

            if (!custom_title) {
                string s = string.Join(" ", title).Replace("\"", "");

                List<byte> zeros = new List<byte>();
                for (int i = 32 - s.Length; i > 0; i--) {
                    zeros.Add(0);
                }
                header.AddRange(strToBa(s));
                header.AddRange(zeros);
            }

            if (has_start) {
                imps.AddRange([15,2]);
            }

            if (impscr) {
                imps.InsertRange(0,[15,2]);
                imps.AddRange([15,1]);
            }
        }

        // ------------------- Write to binary file ---------------------

        public void WriteToBIN(string path) {
            if (File.Exists(path)) {
                File.Delete(path);
            }

            BinaryWriter binfile = new BinaryWriter(File.OpenWrite(path));
            binfile.Write(header.ToArray());
            binfile.Write(imps.ToArray());
            binfile.Write(inst.ToArray());
            binfile.Close();
        }
    }

    class CompilerProgram {
        static Compiler compiler = new Compiler();

        public byte Main(string[] args) { // exe.exe test.asm app.cae
            string infile = "test.asm";
            string outfile = "app.cae";

            try {
                if (args.Length >= 1) infile = args[0];
                if (args.Length >= 2) outfile = args[1];
                if (args.Length >= 3) Console.WriteLine("Too many arguments, ignoring leftover.");
            } catch (Exception ex) { Console.WriteLine(ex.Message); return 1; }

            try {
                compiler.LoadASM(infile);
                compiler.Compile();
                compiler.CheckFlags();
                compiler.WriteToBIN(outfile);

                return 0;
            } catch (Exception ex) {
                Console.WriteLine(ex.Message);
                if (!compiler.flag_err) Console.WriteLine("Line: " + compiler.cur_line);
                Console.WriteLine("File: " + infile);

                return 1;
            }
        }
    }
}